### PR TITLE
JAMES-3138 recompute quotas task & endpoint

### PR DESCRIPTION
**Pull request by Arsnael**
Following the work of #3275 and #3284

### CODE REVIEW

File: src/site/markdown/server/manage-webadmin.md [hunk 1]
```diff
@@ -1151,6 +1152,37 @@ Response codes:
...
+  "type":"recompute-current-quotas",
+  "processedQuotaRoots": 3,
+  "failedQuotaRoots": ["#private&bob@localhost"]
+}
+```
+
+**WARNING**: this task do not take into account concurrent modifications upon a single current quota recomputation. 
+Rerunning the task will *eventually* provide the consistent result. As such we advise to run this task offline. 
+
+In order to ensure being offline, stop the traffic on SMTP, JMAP and IMAP ports, for example via re-configuration or
```
Comment (Arsnael) [inline src/site/markdown/server/manage-webadmin.md:1184]: From Benoit in #3284 :

> We don't alter source of truth we don't need to be offline.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/task/RecomputeCurrentQuotasTask.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Override
+    public Optional<TaskExecutionDetails.AdditionalInformation> details() {
+        Snapshot snapshot = context.snapshot();
+
+        return Optional.of(new Details(Clock.systemUTC().instant(),
+            snapshot.getProcessedQuotaRootCount(),
+            snapshot.getFailedQuotaRoots()
+                .stream()
+                .map(QuotaRoot::toString)
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/task/RecomputeCurrentQuotasTask.java:94]: We should avoid relying on toString. Can we have a asString method?

File: server/container/guice/protocols/webadmin-mailbox/src/main/java/org/apache/james/modules/server/MailboxRoutesModule.java [hunk 3]
```diff
@@ -58,5 +59,9 @@ protected void configure() {
         Multibinder.newSetBinder(binder(), TaskRegistration.class, Names.named(ALL_MAILBOXES_TASKS));
         Multibinder.newSetBinder(binder(), TaskRegistration.class, Names.named(ONE_MAILBOX_TASKS));
         Multibinder.newSetBinder(binder(), TaskRegistration.class, Names.named(ONE_MAIL_TASKS));
+
+        Multibinder.newSetBinder(binder(), TaskRegistration.class, Names.named(USER_QUOTAS_OPERATIONS_INJECTION_KEY))
```
Comment (chibenwa) [inline server/container/guice/protocols/webadmin-mailbox/src/main/java/org/apache/james/modules/server/MailboxRoutesModule.java:63]: Memory in theory won't support current quota recomputation. Thus we need the taskToRequest binding to be done in a separate module, being binded only in products that supports it.

**Comment by Arsnael** [comment]
> Please contribute webadmin documentation

It is already contributed???

File: src/site/markdown/server/manage-webadmin.md [hunk 4]
```diff
@@ -1151,6 +1152,34 @@ Response codes:
 
  - 200: List of users had successfully been returned.
  - 400: Validation issues with parameters
+ 
+### Recomputing current quotas for users
+
+This task is available on top of Guice products.
```
Comment (chibenwa) [inline src/site/markdown/server/manage-webadmin.md:1158]: WXebadmin is only available on top of guice product, this totology can be removed.

Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md:1158]: Not available for memory product I bet

File: server/protocols/webadmin/webadmin-mailbox/pom.xml [hunk 5]
```diff
@@ -68,6 +68,10 @@
             <scope>test</scope>
             <type>test-jar</type>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-mailbox-store</artifactId>
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/pom.xml:73]: Looks very suspicious to me.

**Comment by chibenwa** [comment]
> Could you tell me what would be the use case for others mailbox implementations?

Unreliable InVMEventBus in the face is hard shut downed.

At least once event delivery messing things up. Current quota is not indempotent.

Failing JPA DB writes out of the mailboxManager transactions.

The failure cases are many as you often says ;-)

Why would we prevent other backends to benefit from @Arsnael nice work? Why would we develop 10 times the same thing?

File: server/protocols/webadmin/webadmin-mailbox/pom.xml (hunk 5 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/pom.xml:73]: @Arsnael plans to do a `mailbox-tools-quota` maven package that would replace this awefull mailbox-store dependency.

This is a proposal that I personally support.

**Comment by chibenwa** [comment]
`MemorySpamAssassinContractTest.expungingSpamMessageShouldNotImpactSpamAssassinState » ConditionTimeout`

test this please

**Comment by Arsnael** [comment]
> The JIRA issue doesn't have a definition of done so it's hard to judge if this PR is reaching its goal.

Sorry about that. Just updated it with a DoD and also the link to the proper ADR regarding this task

**Comment by Arsnael** [comment]
> However, I doubt we need to recompute mailbox quotas for any other mailbox backend other than Cassandra.
> Could you tell me what would be the use case for others mailbox implementations?

For the reasons addressed by @chibenwa above, and the fact that the corresponding ADR (https://github.com/apache/james-project/blob/master/src/adr/0028-Recompute-mailbox-quotas.md), that we all approved and agreed to, is stating this as well:

```
This way, non-Cassandra implementation will also benefit from this task.
```

File: server/protocols/webadmin/webadmin-mailbox/pom.xml (hunk 5 above)
Comment (Arsnael) [reply server/protocols/webadmin/webadmin-mailbox/pom.xml]: Done

File: mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/task/JPARecomputeCurrentQuotasServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+import org.apache.james.mailbox.tools.quotas.RecomputeCurrentQuotasServiceContract;
+import org.apache.james.user.api.UsersRepository;
+import org.apache.james.user.jpa.JPAUsersRepository;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+
+import com.google.common.collect.ImmutableList;
+
+@Disabled("JAMES-3138 Does not work, if someone wants to help...")
```
Comment (chibenwa) [inline mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/task/JPARecomputeCurrentQuotasServiceTest.java:49]: Let's not add it in the commit then, and let's not expose the task over webadmin for JPA.

Let's wait a kind contributor to fix it.

File: mailbox/spring/src/main/resources/META-INF/spring/spring-mailbox.xml [hunk 7]
```diff
@@ -39,7 +39,7 @@
     <!-- 
       Mailbox Copier
      -->
-    <bean id="mailboxcopier" class="org.apache.james.mailbox.tools.copier.MailboxCopierImpl"/>
+    <bean id="mailboxcopier" class="org.apache.james.mailbox.tools.quotas.MailboxCopierImpl"/>
```
Comment (chibenwa) [inline mailbox/spring/src/main/resources/META-INF/spring/spring-mailbox.xml:42]: oups?

File: mailbox/store/pom.xml [hunk 8]
```diff
@@ -62,11 +62,25 @@
...
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-data-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-task-json</artifactId>
+        </dependency>
```
Comment (chibenwa) [inline mailbox/store/pom.xml:83]: Why these dependency changes? Still needed?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/CurrentQuotaCalculator.java [hunk 9]
```diff
@@ -26,8 +26,11 @@
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/CurrentQuotaCalculator.java:26]: This class deserve to be in tools/quotas ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/ListeningCurrentQuotaUpdater.java [hunk 10]
```diff
@@ -19,20 +19,25 @@
 package org.apache.james.mailbox.store.quota;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/ListeningCurrentQuotaUpdater.java:19]: This class deserve to be in tools/quotas ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/StoreCurrentQuotaManager.java [hunk 11]
```diff
@@ -20,13 +20,26 @@
 package org.apache.james.mailbox.store.quota;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/StoreCurrentQuotaManager.java:20]: This class deserve to be in tools/quotas ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerQuotaTest.java [hunk 12]
```diff
@@ -37,6 +37,7 @@
 import org.apache.james.mailbox.quota.CurrentQuotaManager;
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerQuotaTest.java:37]: This class deserve to be in tools/quotas ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/quota/ListeningCurrentQuotaUpdaterTest.java [hunk 13]
```diff
@@ -45,6 +45,7 @@
 import org.apache.james.mailbox.events.MailboxListener;
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/quota/ListeningCurrentQuotaUpdaterTest.java:45]: This class deserve to be in tools/quotas ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/quota/StoreCurrentQuotaManagerTest.java [hunk 14]
```diff
@@ -20,18 +20,24 @@
 package org.apache.james.mailbox.store.quota;
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/quota/StoreCurrentQuotaManagerTest.java:20]: This class deserve to be in tools/quotas ?

File: mailbox/tools/copier/src/main/java/org/apache/james/mailbox/tools/quotas/MailboxCopierImpl.java [hunk 15]
```diff
@@ -16,7 +16,7 @@
  * specific language governing permissions and limitations      *
  * under the License.                                           *
  ****************************************************************/
-package org.apache.james.mailbox.tools.copier;
+package org.apache.james.mailbox.tools.quotas;
```
Comment (chibenwa) [inline mailbox/tools/copier/src/main/java/org/apache/james/mailbox/tools/quotas/MailboxCopierImpl.java:19]: OUPS?

File: mailbox/tools/copier/src/test/java/org/apache/james/mailbox/tools/quotas/MailboxCopierTest.java [hunk 16]
```diff
@@ -16,7 +16,7 @@
  * specific language governing permissions and limitations      *
  * under the License.                                           *
  ****************************************************************/
-package org.apache.james.mailbox.tools.copier;
+package org.apache.james.mailbox.tools.quotas;
```
Comment (chibenwa) [inline mailbox/tools/copier/src/test/java/org/apache/james/mailbox/tools/quotas/MailboxCopierTest.java:19]: OUPS?

File: mailbox/tools/quotas/pom.xml [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    <name>Apache James :: Mailbox :: Tools :: Quotas</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-mailbox-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-mailbox-store</artifactId>
```
Comment (chibenwa) [inline mailbox/tools/quotas/pom.xml:40]: That's evil. Moving all the store related quota class in here should solve this.

Then we could easily makes store depends on tools/quota

File: server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml [hunk 18]
```diff
@@ -152,7 +152,7 @@
     <import resource="classpath:META-INF/spring/event-system.xml"/>
 
     <!-- Mailbox Copier -->
-    <bean id="mailboxcopier" class="org.apache.james.mailbox.tools.copier.MailboxCopierImpl"/>
+    <bean id="mailboxcopier" class="org.apache.james.mailbox.tools.quotas.MailboxCopierImpl"/>
```
Comment (chibenwa) [inline server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml:155]: OUPS.

I believe that the copier have nothing to do in quotas package.

**Comment by chibenwa** [comment]
Still a bit of mvn module management wizardry to undergo ;-)

**Comment by mbaechler** [comment]
> > However, I doubt we need to recompute mailbox quotas for any other mailbox backend other than Cassandra.
> > Could you tell me what would be the use case for others mailbox implementations?
> 
> For the reasons addressed by @chibenwa above, and the fact that the corresponding ADR (https://github.com/apache/james-project/blob/master/src/adr/0028-Recompute-mailbox-quotas.md), that we all approved and agreed to, is stating this as well:
> 
> ```
> This way, non-Cassandra implementation will also benefit from this task.
> ```

Thank you for the links and explanations.

File: mailbox/store/pom.xml (hunk 8 above)
Comment (Arsnael) [reply mailbox/store/pom.xml:83]: I need to recheck you are right

File: server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml (hunk 18 above)
Comment (Arsnael) [reply server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml:155]: Sorry not been careful with Intellij obviously...

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraCurrentQuotaManager.java [hunk 19]
```diff
@@ -47,58 +51,70 @@
...
-        this.getCurrentMessageCountStatement = session.prepare(select(CassandraCurrentQuota.MESSAGE_COUNT)
-            .from(CassandraCurrentQuota.TABLE_NAME)
+        this.getCurrentMessageCountStatement = session.prepare(select(MESSAGE_COUNT)
+            .from(TABLE_NAME)
             .where(eq(CassandraCurrentQuota.QUOTA_ROOT, bindMarker())));
-        this.getCurrentStorageStatement = session.prepare(select(CassandraCurrentQuota.STORAGE)
-            .from(CassandraCurrentQuota.TABLE_NAME)
+        this.getCurrentStorageStatement = session.prepare(select(STORAGE)
+            .from(TABLE_NAME)
+            .where(eq(CassandraCurrentQuota.QUOTA_ROOT, bindMarker())));
```
Comment (nvduc91) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraCurrentQuotaManager.java:72]: Small thing but `CassandraCurrentQuota.QUOTA_ROOT` can import static

Comment (Arsnael) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraCurrentQuotaManager.java:72]: I think I had it before, seems got lost in the multiples rebases... Thanks for spotting it !

File: mailbox/tools/quotas/pom.xml (hunk 17 above)
Comment (Arsnael) [reply mailbox/tools/quotas/pom.xml:40]: Ok problem is, among the classes you asked me to move there... :
* CurrentQuotaCalculator: using `MailboxSessionMapperFactory`, from mailbox-store
* AbstractMessageIdManagerQuotaTest: using `MessageIdManagerTestSystem` from mailbox-store

And if we don't move those... well I still need `CurrentQuotaCalculator` in the service. So hard to see how I can remove that dependency without an other yet refactoring, and tbh tools-indexer is having evil too (didn't say was right though :) )

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerQuotaTest.java (hunk 12 above)
Comment (Arsnael) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerQuotaTest.java:37]: Not sure about this one?

Comment (Arsnael) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerQuotaTest.java:37]: Not sure about this one?

**Comment by Arsnael** [comment]
Rebasing on #3284 (again). Reworked the PR to add the wrapping task in the new `apache-james-mailbox-tools-quota-recompute` that we agreed on yesterday, introduced by #3284

File: mailbox/tools/quota-recompute/pom.xml [hunk 20]
```diff
@@ -36,15 +36,35 @@
...
+            <scope>test</scope>
+            <type>test-jar</type>
+        </dependency>
         <dependency>
             <groupId>${james.groupId}</groupId>
             <artifactId>james-server-data-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-task-json</artifactId>
```
Comment (mbaechler) [inline mailbox/tools/quota-recompute/pom.xml:51]: importing json into mailbox-api?

**Comment by Arsnael** [comment]
```
[7f22e8dc55b093110c5264175aa90d2ace930274] [ERROR] Errors: 
[7f22e8dc55b093110c5264175aa90d2ace930274] [ERROR]   CassandraDumbBlobStoreCacheTest.shouldReturnDataWhenCacheSmallDataInConfigurationTTL » ConditionTimeout
```
hmmmmm... test this please

File: mailbox/tools/quota-recompute/pom.xml (hunk 20 above)
Comment (Arsnael) [reply mailbox/tools/quota-recompute/pom.xml:51]: That's not `mailbox-api` but `mailbox-tools-quota-recompute` module here?
Well to create the task `RecomputeCurrentQuotasTask` wrapping around the service `RecomputeCurrentQuotasService`, with its DTOs that need `james-server-task-json`? `RecomputeCurrentQuotasTaskDTO` and `RecomputeCurrentQuotasTaskAdditionalInformationDTO`)

Can you please then be more explicit in why you think it's wrong again, and what are your suggestions around that? Thanks.

**Comment by Arsnael** [comment]
Rebased with master

File: mailbox/tools/quota-recompute/pom.xml (hunk 20 above)
Comment (mbaechler) [reply mailbox/tools/quota-recompute/pom.xml:51]: I checked twice before adding this comment: github displayed mailbox-api ... now it's no longer the case.

Comment (Arsnael) [reply mailbox/tools/quota-recompute/pom.xml:51]: There is some serious issues with github seems these days...

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/QuotaRoot.java [hunk 21]
```diff
@@ -73,4 +73,9 @@ public String toString() {
                 .add("domain", domain)
                 .toString();
     }
+
+    public String asString() {
+        return domain.map(domainValue -> value + "@" + domainValue.asString())
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/QuotaRoot.java:78]: Not sure to get your point. I think as I understood QuotaRoot, the `value` is pretty much the composition of:
```
USER_NAMESPACE (like "#private" for example) + SEPARATOR ("#") + username
```
So for the user bob I guess the value would be: `#private&bob`

Then you have the optional `domain` (like james.org or whatever)

The username seems already in value, so I don't get your point here...?

**Comment by Arsnael** [comment]
The tests are green but the build crashed in the dockering phase

test this please

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/QuotaRoot.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/QuotaRoot.java:78]: Well, you may be right. Honestly, the only thing I see is `String` so I don't know what's in.
May having a `QuotaRootTest` class that covers `asString` would help?

**Comment by chibenwa** [comment]
> Do we have integration tests?

We need at least a test showing that the task is well exposed.

**Comment by Arsnael** [comment]
> > Do we have integration tests?
> 
> We need at least a test showing that the task is well exposed.

There is one in rabbitmq integration tests (distributed james). But I guess you mean I need for other products as well?

**Comment by chibenwa** [comment]
> There is one in rabbitmq integration tests (distributed james). But I guess you mean I need for other products as well?

That's enough for me

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/QuotaRootTest.java [hunk 22]
```diff
@@ -28,4 +33,23 @@
...
+        assertThat(quotaRoot.asString()).isEqualTo(value);
+    }
+
+    @Test
+    void asStringShouldReturnValueWithDomainWhenHasDomain() {
+        String value = "#private&bob";
+        Domain domain = Domain.of("apache.org");
+        QuotaRoot quotaRoot = QuotaRoot.quotaRoot(value, Optional.of(domain));
+
+        String expectedValue = value + "@" + domain.asString();
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/QuotaRootTest.java:51]: please don't use operation to compose expected result ... you should hardcode it in this case

**Comment by mbaechler** [comment]
> > > Do we have integration tests?
> > 
> > 
> > We need at least a test showing that the task is well exposed.
> 
> There is one in rabbitmq integration tests (distributed james). But I guess you mean I need for other products as well?

This test computes nothing or I miss something?

**Comment by chibenwa** [comment]
> This test computes nothing or I miss something?

We're not doing for other consistency tasks.

Fault injection in integration tests requires more work as they don't rely on testing session.

I propose handling this as a separate PR enhancement.

**Comment by chibenwa** [comment]
@mbaechler see https://github.com/linagora/james-project/pull/3329

**Comment by Arsnael** [comment]
> I propose handling this as a separate PR enhancement.

cf #3329

**Comment by mbaechler** [comment]
> > This test computes nothing or I miss something?
> 
> We're not doing for other consistency tasks.
> 
> Fault injection in integration tests requires more work as they don't rely on testing session.
> 
> I propose handling this as a separate PR enhancement.

I don't request to test inconsistencies but the run on a non-empty server.

**Comment by chibenwa** [comment]
> I don't request to test inconsistencies but the run on a non-empty server.

In the end, that's what https://github.com/linagora/james-project/pull/3329 provide (and even more!)

**Comment by mbaechler** [comment]
> > I don't request to test inconsistencies but the run on a non-empty server.
> 
> In the end, that's what #3329 provide (and even more!)

Do you have a fix-nothing integration test for non-empty server?

**Comment by chibenwa** [comment]
> Do you have a fix-nothing integration test for non-empty server?

For sure, we always can write more tests.

**Comment by Arsnael** [comment]
```
[17a9ac75100b94377a12093b60d65ecb1264926a] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project cassandra-jmap-draft-integration-testing: There was a timeout or other error in the fork -> [Help 1]
```
test this please

**Comment by Arsnael** [comment]
> Do you have a fix-nothing integration test for non-empty server?

@mbaechler : This is a remark for #3329 right? If there is no other remarks for this PR, is it possible to approve, or do you have other concerns?

**Comment by chibenwa** [comment]
@mbaechler see https://github.com/linagora/james-project/pull/3329/commits/ff0d46c5bf6d68754809609ea30ae913713fb082

test this please

**Comment by mbaechler** [comment]
> > Do you have a fix-nothing integration test for non-empty server?
> 
> @mbaechler : This is a remark for #3329 right? If there is no other remarks for this PR, is it possible to approve, or do you have other concerns?

Well, you decided to split things in two PR, that's ok, but please don't blame me if my comments at the origin of this split is in the wrong place.

My point was: I didn't expect the tests @chibenwa wrote, I just wanted an integration test actually demonstrating some things run, in this PR.

> @mbaechler see ff0d46c

Thanks. I'll approve this PR.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**